File: data_structures/marketing_events.py
import random


class MarketingCampaign:
    """Base class for marketing campaigns affecting the DAO."""

    def __init__(self, dao, budget: float = 50):
        self.dao = dao
        self.budget = budget

    def execute(self, sim):
        raise NotImplementedError


class DemandBoostCampaign(MarketingCampaign):
    """Increase token price by spending treasury funds on promotion."""

    def __init__(self, dao, budget: float = 50, price_boost: float = 0.1):
        super().__init__(dao, budget)
        self.price_boost = price_boost

    def execute(self, sim):
        spent = self.dao.treasury.withdraw("DAO_TOKEN", self.budget)
        old_price = self.dao.treasury.get_token_price("DAO_TOKEN")
        new_price = old_price * (1 + self.price_boost)
        self.dao.treasury.update_token_price("DAO_TOKEN", new_price)
        if self.dao.event_bus:
            self.dao.event_bus.publish(
                "marketing_campaign",
                step=self.dao.current_step,
                type="demand_boost",
                budget=spent,
                old_price=old_price,
                new_price=new_price,
            )


class RecruitmentCampaign(MarketingCampaign):
    """Attract new members to the DAO using marketing funds."""

    def __init__(self, dao, budget: float = 50, recruits: int = 2):
        super().__init__(dao, budget)
        self.recruits = recruits

    def execute(self, sim):
        from agents.passive_member import PassiveMember

        spent = self.dao.treasury.withdraw("DAO_TOKEN", self.budget)
        price = self.dao.treasury.get_token_price("DAO_TOKEN")
        new_ids = []
        for i in range(self.recruits):
            uid = f"Recruit_{sim.schedule.steps}_{i}"
            member = PassiveMember(uid, model=self.dao, tokens=100, reputation=0, location="global")
            self.dao.add_member(member)
            sim.schedule.add(member)
            new_ids.append(uid)
        if self.dao.event_bus:
            self.dao.event_bus.publish(
                "marketing_campaign",
                step=self.dao.current_step,
                type="recruitment",
                budget=spent,
                new_members=new_ids,
                old_price=price,
                new_price=price,
            )


class SocialMediaCampaign(MarketingCampaign):
    """Run a small social media push to slightly boost token demand."""

    def __init__(self, dao, budget: float = 20, price_boost: float = 0.05):
        super().__init__(dao, budget)
        self.price_boost = price_boost

    def execute(self, sim):
        spent = self.dao.treasury.withdraw("DAO_TOKEN", self.budget)
        old_price = self.dao.treasury.get_token_price("DAO_TOKEN")
        new_price = old_price * (1 + self.price_boost)
        self.dao.treasury.update_token_price("DAO_TOKEN", new_price)
        if self.dao.event_bus:
            self.dao.event_bus.publish(
                "marketing_campaign",
                step=self.dao.current_step,
                type="social_media",
                budget=spent,
                old_price=old_price,
                new_price=new_price,
            )


class ReferralBonusCampaign(MarketingCampaign):
    """Reward existing members for recruiting friends."""

    def __init__(self, dao, budget: float = 60, recruits: int = 1, bonus: float = 10):
        super().__init__(dao, budget)
        self.recruits = recruits
        self.bonus = bonus

    def execute(self, sim):
        from agents.passive_member import PassiveMember

        spent = self.dao.treasury.withdraw("DAO_TOKEN", self.budget)
        price = self.dao.treasury.get_token_price("DAO_TOKEN")
        new_ids = []
        referrer = random.choice(self.dao.members) if self.dao.members else None
        if referrer:
            referrer.tokens += self.bonus
        for i in range(self.recruits):
            uid = f"Referral_{sim.schedule.steps}_{i}"
            member = PassiveMember(uid, model=self.dao, tokens=100, reputation=0, location="global")
            self.dao.add_member(member)
            sim.schedule.add(member)
            new_ids.append(uid)
        if self.dao.event_bus:
            self.dao.event_bus.publish(
                "marketing_campaign",
                step=self.dao.current_step,
                type="referral_bonus",
                budget=spent,
                new_members=new_ids,
                referrer=getattr(referrer, "unique_id", None),
                old_price=price,
                new_price=price,
            )

File: lib/data-structures/marketing-events.ts
// Marketing Campaign Events - various marketing strategies for DAOs
// Port from data_structures/marketing_events.py

import type { DAO } from './dao';
import type { DAOModel } from '../engine/model';
import { PassiveMember } from '../agents/passive-member';
import { randomChoice } from '../utils/random';

/**
 * Base class for marketing campaigns
 */
export abstract class MarketingCampaign {
  dao: DAO;
  budget: number;

  constructor(dao: DAO, budget: number = 50) {
    this.dao = dao;
    this.budget = budget;
  }

  abstract execute(sim: DAOModel): void;
}

/**
 * Demand Boost Campaign - increase token price by spending treasury funds
 */
export class DemandBoostCampaign extends MarketingCampaign {
  priceBoost: number;

  constructor(dao: DAO, budget: number = 50, priceBoost: number = 0.1) {
    super(dao, budget);
    this.priceBoost = priceBoost;
  }

  execute(sim: DAOModel): void {
    const spent = this.dao.treasury.withdraw(
      'DAO_TOKEN',
      this.budget,
      sim.currentStep
    );

    const oldPrice = this.dao.treasury.getTokenPrice('DAO_TOKEN');
    const newPrice = oldPrice * (1 + this.priceBoost);
    this.dao.treasury.updateTokenPrice('DAO_TOKEN', newPrice);

    if (this.dao.eventBus) {
      this.dao.eventBus.publish('marketing_campaign', {
        step: sim.currentStep,
        type: 'demand_boost',
        budget: spent,
        oldPrice,
        newPrice,
      });
    }
  }
}

/**
 * Recruitment Campaign - attract new members to the DAO
 */
export class RecruitmentCampaign extends MarketingCampaign {
  recruits: number;

  constructor(dao: DAO, budget: number = 50, recruits: number = 2) {
    super(dao, budget);
    this.recruits = recruits;
  }

  execute(sim: DAOModel): void {
    const spent = this.dao.treasury.withdraw(
      'DAO_TOKEN',
      this.budget,
      sim.currentStep
    );

    const price = this.dao.treasury.getTokenPrice('DAO_TOKEN');
    const newIds: string[] = [];

    for (let i = 0; i < this.recruits; i++) {
      const uid = `Recruit_${sim.currentStep}_${i}`;
      const member = new PassiveMember(
        uid,
        sim,
        100,
        0,
        'global'
      );

      this.dao.addMember(member);
      sim.schedule.add(member);
      newIds.push(uid);
    }

    if (this.dao.eventBus) {
      this.dao.eventBus.publish('marketing_campaign', {
        step: sim.currentStep,
        type: 'recruitment',
        budget: spent,
        newMembers: newIds,
        oldPrice: price,
        newPrice: price,
      });
    }
  }
}

/**
 * Social Media Campaign - small social media push to boost token demand
 */
export class SocialMediaCampaign extends MarketingCampaign {
  priceBoost: number;

  constructor(dao: DAO, budget: number = 20, priceBoost: number = 0.05) {
    super(dao, budget);
    this.priceBoost = priceBoost;
  }

  execute(sim: DAOModel): void {
    const spent = this.dao.treasury.withdraw(
      'DAO_TOKEN',
      this.budget,
      sim.currentStep
    );

    const oldPrice = this.dao.treasury.getTokenPrice('DAO_TOKEN');
    const newPrice = oldPrice * (1 + this.priceBoost);
    this.dao.treasury.updateTokenPrice('DAO_TOKEN', newPrice);

    if (this.dao.eventBus) {
      this.dao.eventBus.publish('marketing_campaign', {
        step: sim.currentStep,
        type: 'social_media',
        budget: spent,
        oldPrice,
        newPrice,
      });
    }
  }
}

/**
 * Referral Bonus Campaign - reward existing members for recruiting friends
 */
export class ReferralBonusCampaign extends MarketingCampaign {
  recruits: number;
  bonus: number;

  constructor(
    dao: DAO,
    budget: number = 60,
    recruits: number = 1,
    bonus: number = 10
  ) {
    super(dao, budget);
    this.recruits = recruits;
    this.bonus = bonus;
  }

  execute(sim: DAOModel): void {
    const spent = this.dao.treasury.withdraw(
      'DAO_TOKEN',
      this.budget,
      sim.currentStep
    );

    const price = this.dao.treasury.getTokenPrice('DAO_TOKEN');
    const newIds: string[] = [];

    // Choose a random referrer
    const referrer =
      this.dao.members.length > 0
        ? randomChoice(this.dao.members)
        : null;

    if (referrer) {
      referrer.tokens += this.bonus;
    }

    // Add new members
    for (let i = 0; i < this.recruits; i++) {
      const uid = `Referral_${sim.currentStep}_${i}`;
      const member = new PassiveMember(
        uid,
        sim,
        100,
        0,
        'global'
      );

      this.dao.addMember(member);
      sim.schedule.add(member);
      newIds.push(uid);
    }

    if (this.dao.eventBus) {
      this.dao.eventBus.publish('marketing_campaign', {
        step: sim.currentStep,
        type: 'referral_bonus',
        budget: spent,
        newMembers: newIds,
        referrer: referrer?.uniqueId || null,
        oldPrice: price,
        newPrice: price,
      });
    }
  }
}
